### PR TITLE
[fpmsyncd] Implement pending route suppression feature

### PR DESCRIPTION
**Pull request by stepanblyschak**
**DEPENDS**: https://github.com/sonic-net/sonic-swss/pull/2512

**What I did**

I implemented support to enable pending routes suppression feature. When this feature is enabled, fpmsyncd will wait for reply from orchagent before sending offload status message to zebra.

**Why I did it**

This is done to not announce routes which aren't yet offloaded in HW.

**How I verified it**

UT and manual tests.

**Details if related**

### CODE REVIEW

File: fpmsyncd/routesync.cpp
Comment (venkatmahalingam) [inline fpmsyncd/routesync.cpp:834]: Hope you have a test case to test this MPLS label route.

File: fpmsyncd/routesync.cpp
Comment (venkatmahalingam) [inline fpmsyncd/routesync.cpp:1362]: Dont we get Del operation in the event? why should we assume this?

File: fpmsyncd/routesync.cpp
Comment (venkatmahalingam) [inline fpmsyncd/routesync.cpp:1414]: When there are multiple route updates for the given route, how do we make sure that the response to sent for the right route-update?

Comment (stepanblyschak) [reply fpmsyncd/routesync.cpp:1416]: The constructed RTM_NEWROUTE message contains prefix + vrf + protocol of the route

File: fpmsyncd/routesync.cpp
Comment (stepanblyschak) [reply fpmsyncd/routesync.cpp:1364]: If we get an response for a delete operation we don't get original FVs. We use this knowledge to filter out DEL events we are not interested in.

ResponsePublisher for reference: https://github.com/sonic-net/sonic-swss/blob/master/orchagent/response_publisher.cpp#L106

File: fpmsyncd/routesync.cpp
Comment (stepanblyschak) [reply fpmsyncd/routesync.cpp:834]: I don't add additional flows for MPLS route handling. So it should be covered by existing MPLS route tests.

**Comment by prsunny** [comment]
@stepanblyschak , can you please test for backward compatibility?

**Comment by stepanblyschak** [comment]
@prsunny It can be merged without https://github.com/sonic-net/sonic-buildimage/pull/12853. BGP facts and FIB test passed on T1-64-LAG with this change.

Please note, that I had to mark VS tests with ```xfail``` because they require VS changes done in https://github.com/sonic-net/sonic-buildimage/pull/12853. Otherwise, it is not possible to merge one without the other in order to not break tests or image.
Later this mark has to be removed once sonic-buildimage PR is merged.

File: fpmsyncd/fpmlink.cpp
Comment (StormLiangMS) [inline fpmsyncd/fpmlink.cpp:295]: wording issue, then -> than

**Comment by StormLiangMS** [comment]
@jcaiMR could you help to review?

File: fpmsyncd/fpmsyncd.cpp
Comment (StormLiangMS) [inline fpmsyncd/fpmsyncd.cpp:76]: @stepanblyschak what's the usage for RTM_NETLINK?

File: fpmsyncd/fpmsyncd.cpp
Comment (StormLiangMS) [inline fpmsyncd/fpmsyncd.cpp:260]: @stepanblyschak when do the disable, is that possible some in flight messages are lost?

File: fpmsyncd/fpmlink.cpp
Comment (stepanblyschak) [reply fpmsyncd/fpmlink.cpp]: Thanks, fixed

File: fpmsyncd/fpmsyncd.cpp
Comment (stepanblyschak) [reply fpmsyncd/fpmsyncd.cpp:260]: Once disable fpmsyncd will reply imidiatelly on receiving route from FRR, no messages are lost

Comment (StormLiangMS) [reply fpmsyncd/fpmsyncd.cpp:260]: @stepanblyschak 
Assume the route A sent to the orchagent to offload, fpmsyncd was waiting for the response. At this time, the route A is in hold status, if we disable the suppress, fpmsyncd won't care the msg sent back to announce route A is offloaded to HW successfully, right? In that case, the route A is offloaded to HW, but the status in FRR is suppressed, right?

Comment (stepanblyschak) [reply fpmsyncd/fpmsyncd.cpp:260]: @StormLiangMS Oh, I get it. Right, we will have suppressed route in FRR in case user disables prior to processing update from orchagent. That can be one or many routes.
We could listen to response channel regardless of suppress status and send back offload flag again but that would generate twice as much traffic to zebra in case feature is disabled.

How about in case of transition from enabled -> disabled fetch all routes present in APPL_DB (it will include those routes that are not set offloaded yet) and send offload status?

Comment (StormLiangMS) [reply fpmsyncd/fpmsyncd.cpp:267]: @stepanblyschak 
The other way we may have a snapshot for the routes in hold in FRR when user disable it, and keep listen to response channel and only send back offload flag which matched with this snapshot list?

In order to avoid the race condition for when we take the snapshot, we may allow a little bit overlap time, during that time frame, send twice traffic to Zebra.

Comment (StormLiangMS) [reply fpmsyncd/fpmsyncd.cpp:267]: > @stepanblyschak The other way we may have a snapshot for the routes in hold in FRR when user disable it, and keep listen to response channel and only send back offload flag which matched with this snapshot list?
> 
> In order to avoid the race condition for when we take the snapshot, we may allow a little bit overlap time, during that time frame, send twice traffic to Zebra.

I rethink it, maybe to make it simple, your idea is good enough, our purpose here it to make sure when suppress enabled, to make sure HW offload happened before advertised out, but if this is disabled, just make sure everything is advertised out would be ok, no need to make sure offload happened before advertise at this time.

**Comment by prsunny** [comment]
@StormLiangMS , are we good to merge this? Can you please sign-off if no further comments?

File: fpmsyncd/fpmsyncd.cpp
Comment (StormLiangMS) [reply fpmsyncd/fpmsyncd.cpp:267]: @stepanblyschak should we revert the sequence, first, setSuppressionEnabled to disable the offload process, then do the markRoutesOffloaded? Otherwise it still timing that some inflight routes after the markRoutesOffloaded.

**Review by StormLiangMS — APPROVED** [review]
LGTM

**Comment by StormLiangMS** [comment]
@prsunny this looks good to me.

**Comment by StormLiangMS** [comment]
@stepanblyschak could you also have the testcase in sonic-mgmt to cover this FIB suppress feature, and pls involve Jcai and me for the code review.

**Comment by stepanblyschak** [comment]
@StormLiangMS @jcaiMR BGP suppress FIB pending test cases are already available in PRs - test plan + implementation:
- https://github.com/sonic-net/sonic-mgmt/pull/7475
- https://github.com/sonic-net/sonic-mgmt/pull/7430

**Pull request merged**
